File: NC/find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:

        def findMin(start: int, end: int):

            if start == end:
                return nums[start]

            if start == end-1:
                return min(nums[start], nums[end])

            mid = int((start+end)/2)

            if nums[start] > nums[mid]:
                return findMin(start, mid)

            elif nums[mid] > nums[end]:
                return findMin(mid, end)

            return nums[start]

        return findMin(0, len(nums)-1)

File: NC/find-minimum-in-rotated-sorted-array.cpp
class Solution
{
public:
    int findMin(vector<int> &nums, int start, int end)
    {
        if (start == end)
        {
            return nums[start];
        }

        // Sorted scenario
        if (nums[start] < nums[end])
        {
            return nums[start];
        }

        int mid = (start + end) / 2;

        return min(findMin(nums, start, mid), findMin(nums, mid + 1, end));
    }

    int findMin(vector<int> &nums)
    {
        return findMin(nums, 0, nums.size() - 1);
    }
};
